### PR TITLE
Cleanup NativeAOT math helpers

### PR DESCRIPTION
**Pull request by MichalPetryka**
Extracted from #98858.

cc @jkotas I'd like the cleanup here to be merged here before moving the Checked multiply helpers around.

### CODE REVIEW

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs [hunk 1]
```diff
@@ -2,381 +2,329 @@
...
+            uint val1High = High32Bits((ulong)left);
+            uint val2High = High32Bits((ulong)right);
 
             ulong valMid;
 
             if (val1High == 0)
             {
                 // Compute the 'middle' bits of the long multiplication
-                valMid = Mul32x32To64(val2High, (uint)i);
+                valMid = Math.BigMul(val2High, (uint)left);
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs:131]: Switching this to Math.BigMul seems to be a deoptimization when compiling for VEX-enabled baseline. The Bmi2 special path in Math.BigMul is slower than a regular multiplication that's here today:

My usual lame microbenchmark:
```
using System;
using System.Diagnostics;
using System.Runtime.CompilerServices;

for (; ; )
{
    var sw = new Stopwatch();
    sw.Start();
    Work(2_000_000_000);
    Console.WriteLine(sw.ElapsedMilliseconds);
}

[MethodImpl(MethodImplOptions.NoInlining)]
long Work(long x)
{
    long sum = 0;
    for (long i = 1; i < x; i += 10)
        sum += checked(i * i);
    return sum;
}
```

Baseline: ~680ms
PR w/ --instruction-set:bmi2: ~800ms

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs [hunk 2]
```diff
@@ -2,381 +2,329 @@
...
 
-        private const string RuntimeLibrary = "*";
+        [RuntimeExport("Dbl2IntOvf")]
+        public static int Dbl2IntOvf(double value)
+        {
+            // Note that this expression also works properly for val = NaN case
+            if (value is > -IntMaxValueOffset - 1 and < IntMaxValueOffset)
+            {
+                int ret = (int)value;
+                // since no overflow can occur, the value always has to be within 1
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs:28]: I know we have these asserts in the unmanaged impl currently, but I do not think they provide any value. The Intel PR is deleting them.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs (hunk 1 above)
Comment (MichalPetryka) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs:131]: cc @tannergooding do we want to remove the BMI2 path there in this case?

Comment (MichalPetryka) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs:131]: It actually seems like `SkipLocalsInit` is making the JIT confused in the BMI2 path:
```asm
G_M56079_IG01:        ; bbWeight=1, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, nogc <-- Prolog IG
       56           push     esi
       83EC08       sub      esp, 8
						;; size=4 bbWeight=1 PerfScore 1.25
G_M56079_IG02:        ; bbWeight=1, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref
                    ; INLRT @ 0x003[E-]
       8B0424       mov      eax, dword ptr [esp]
       23D0         and      edx, eax
       89542404     mov      dword ptr [esp+0x04], edx
       8B742404     mov      esi, dword ptr [esp+0x04]
       8BD1         mov      edx, ecx
       C4E26BF6D6   mulx     edx, edx, esi
       83C800       or       eax, 0
       83CA00       or       edx, 0
						;; size=26 bbWeight=1 PerfScore 7.00
G_M56079_IG03:        ; bbWeight=1, epilog, nogc, extend
       83C408       add      esp, 8
       5E           pop      esi
       C3           ret      
						;; size=5 bbWeight=1 PerfScore 1.75
```

Without that:
```asm
G_M56079_IG01:        ; bbWeight=1, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, nogc <-- Prolog IG
						;; size=0 bbWeight=1 PerfScore 0.00
G_M56079_IG02:        ; bbWeight=1, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref
                    ; INLRT @ 0x003[E-]
       33C0         xor      eax, eax
       8BD1         mov      edx, ecx
       C4E26BF6D0   mulx     edx, edx, eax
						;; size=9 bbWeight=1 PerfScore 3.50
G_M56079_IG03:        ; bbWeight=1, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, epilog, nogc
       C3           ret      
						;; size=1 bbWeight=1 PerfScore 1.00
```

Comment (jkotas) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs:131]: FWIW, I think mulx should be an optimization in the JIT that automatically kicks in for all multiplications that fit the pattern. It does not make sense to make everybody out there to tweak their 32-bit*32-bit->64-bit multiplication to deal with the BigMul helper.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs [hunk 3]
```diff
@@ -2,381 +2,313 @@
...
+            uint val1High = High32Bits((ulong)left);
+            uint val2High = High32Bits((ulong)right);
 
             ulong valMid;
 
             if (val1High == 0)
             {
                 // Compute the 'middle' bits of the long multiplication
-                valMid = Mul32x32To64(val2High, (uint)i);
+                valMid = (ulong)val2High * (uint)left;
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs:115]: I think the code was easier to understand with the `Mul32x32To64` helper function. I do not think it is an improvement to inline it like this.

**Review by jkotas — APPROVED** [review]
LGTM. Thank you!

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs [hunk 4]
```diff
@@ -2,381 +2,313 @@
...
 {
     /// <summary>
-    /// Math helpers for generated code. The helpers marked with [RuntimeExport] and the type
-    /// itself need to be public because they constitute a public contract with the .NET Native toolchain.
+    /// Math helpers for generated code. The helpers here are referenced by the runtime.
     /// </summary>
+    [StackTraceHidden]
     internal static partial class MathHelpers
     {
+        private const double IntMaxValueOffset = 2147483648.0; // 2^31, int.MaxValue + 1
```
Comment (tannergooding) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs:18]: nit: Convention is to use the .NET names, not the C# keyword names

This should be `Int32MaxValueOffset` and the next shoul dbe `UInt32MaxValueOffset`. It's generally good to follow the convention even for internal APIs.

**Review by tannergooding — APPROVED** [review]

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs [hunk 5]
```diff
@@ -2,381 +2,313 @@
...
+        private const double UIntMaxValueOffset = 4294967296.0; // 2^32, uint.MaxValue + 1
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs:19]:
```suggestion
        private const double Int32MaxValueOffset = (double)int.MaxValue + 1;
        private const double UInt32MaxValueOffset = (double)uint.MaxValue + 1;
```

**Comment by jkotas** [comment]
The outer loop failures look unrelated.

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs [hunk 6]
```diff
@@ -2,381 +2,313 @@
...
+        }
+
+        [RuntimeExport("Dbl2LngOvf")]
+        public static long Dbl2LngOvf(double value)
+        {
+            const double two63 = Int32MaxValueOffset * UInt32MaxValueOffset;
+
+            // Note that this expression also works properly for val = NaN case
+            // We need to compare with the very next double to two63. 0x402 is epsilon to get us there.
+            if (value is > -two63 - 0x402 and < two63)
```
Comment (MichalPetryka) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs:54]:
```suggestion
            if (value is > Math.BitDecrement(-two63) and < two63)
```
Would this work here?

Comment (hamarb123) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs:54]: Not a compile-time constant

Comment (jkotas) [reply src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs:54]: > Would this work here?

Nope. BitDecrement is not a JIT intrinsic, and it is too complicated to make the JIT inline in and evaluate it into a constant. 

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunGATCAFdgAGxgMAsgAoAlBy49FxAOyTsGABYA6AEIBLDABEYYWPhgA7DFPJaADDIDcC7gF8aroA===

File: src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs [hunk 7]
```diff
@@ -2,381 +2,313 @@
...
 namespace Internal.Runtime.CompilerHelpers
 {
     /// <summary>
-    /// Math helpers for generated code. The helpers marked with [RuntimeExport] and the type
-    /// itself need to be public because they constitute a public contract with the .NET Native toolchain.
+    /// Math helpers for generated code. The helpers here are referenced by the runtime.
     /// </summary>
-    internal static partial class MathHelpers
+    [StackTraceHidden]
+    internal static class MathHelpers
```
Comment (jkotas) [inline src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/MathHelpers.cs:16]:
```suggestion
    internal static partial class MathHelpers
```
Fix build break

**Comment by jkotas** [comment]
Thank you!

**Pull request merged**